hswidget/ds1widget/tile_records.go

```diff
@@ -0,0 +1,207 @@
+
+	newWall := d2ds1.WallRecord{
+		Type:        d2enum.TileType(state.addWallState.tileType),
+		Zero:        byte(state.addWallState.zero),
+		Prop1:       byte(state.addWallState.prop1),
+		Sequence:    byte(state.addWallState.sequence),
+		Unknown1:    byte(state.addWallState.unknown1),
+		Style:       byte(state.addWallState.style),
+		Unknown2:    byte(state.addWallState.unknown2),
+		HiddenBytes: byte(state.addWallState.hidden),
+	}
+
+	return newWall
+}
+
+func (p *DS1Widget) addFloor() {
+	state := p.getState()
+
+	state.addFloorShadowState.cb = func() {
+		newFloor := p.createFloorShadowRecord()
+
+		for y := range p.ds1.Tiles {
+			for x := range p.ds1.Tiles[y] {
+				p.ds1.Tiles[y][x].Floors = append(p.ds1.Tiles[y][x].Floors, newFloor)
+			}
+		}
+
+		p.ds1.NumberOfFloors++
+
+		p.recreateLayerStreamTypes()
+	}
+
+	state.mode = ds1EditorModeAddFloorShadow
+}
+
+func (p *DS1Widget) editFloor() {
+	state := p.getState()
+
+	state.addFloorShadowState.cb = func() {
+		newFloor := p.createFloorShadowRecord()
+
+		p.ds1.Tiles[state.tileY][state.tileY].Floors[state.object] = newFloor
+	}
+	state.mode = ds1EditorModeAddFloorShadow
+}
+
+func (p *DS1Widget) deleteFloorRecord() {
+	state := p.getState()
+
+	for y := range p.ds1.Tiles {
+		for x := range p.ds1.Tiles[y] {
+			newFloors := make([]d2ds1.FloorShadowRecord, 0)
+
+			for n, i := range p.ds1.Tiles[y][x].Floors {
+				if n != int(state.object) {
+					newFloors = append(newFloors, i)
+				}
+			}
+
+			p.ds1.Tiles[y][x].Floors = newFloors
+		}
+	}
+
+	p.ds1.NumberOfFloors--
+	p.recreateLayerStreamTypes()
+}
+
+func (p *DS1Widget) addWall() {
+	state := p.getState()
+
+	state.addWallState.cb = func() {
+		newWall := p.createWallRecord()
+
+		for y := range p.ds1.Tiles {
+			for x := range p.ds1.Tiles[y] {
+				p.ds1.Tiles[y][x].Walls = append(p.ds1.Tiles[y][x].Walls, newWall)
+			}
+		}
+
+		p.ds1.NumberOfWalls++
+
+		p.recreateLayerStreamTypes()
+	}
+
+	state.mode = ds1EditorModeAddWall
+}
+
+func (p *DS1Widget) editWall() {
+	state := p.getState()
+
+	state.addWallState.cb = func() {
+		newWall := p.createWallRecord()
+
+		p.ds1.Tiles[state.tileY][state.tileY].Walls[state.object] = newWall
+	}
+	state.mode = ds1EditorModeAddWall
+}
+
+func (p *DS1Widget) deleteWall() {
+	state := p.getState()
+
+	for y := range p.ds1.Tiles {
+		for x := range p.ds1.Tiles[y] {
+			newWalls := make([]d2ds1.WallRecord, 0)
+
+			for n, i := range p.ds1.Tiles[y][x].Walls {
+				if n != int(state.object) {
+					newWalls = append(newWalls, i)
+				}
+			}
+
+			p.ds1.Tiles[y][x].Walls = newWalls
+		}
+	}
+
+	p.ds1.NumberOfWalls--
+	p.recreateLayerStreamTypes()
+}
+
+func (p *DS1Widget) addShadow() {
+	state := p.getState()
+
+	state.addFloorShadowState.cb = func() {
+		newShadow := p.createFloorShadowRecord()
+
+		for y := range p.ds1.Tiles {
+			for x := range p.ds1.Tiles[y] {
+				p.ds1.Tiles[y][x].Shadows = make([]d2ds1.FloorShadowRecord, 1)
+				p.ds1.Tiles[y][x].Shadows[0] = newShadow
+			}
+		}
+	}
```

Comment:
is this setting the `Shadows` for all tiles to the `newShadow` record when you add a new `shadow`... ?

Reply:
what if the tile already had a shadow...?

Reply:
- when one tile has a shadow, then another tiles too (`DS1.LayerStreamTypes`)
- it is only called, when tile hasn't any shadow

Reply:
Ok, after looking at the tile record struct, i think i get it; tiles have a variable number of walls, floors, and shadows.